spec/client/ics-009-loopback-cilent/README.md

```diff
@@ -0,0 +1,247 @@
+}
+```
+
+### Header
+
+The header includes the identifier of the light client and the current block height of the local ledger.
+
+```typescript
+interface Header {
+  identifier: string
+  height: Height
+}
+```
+
+Header implements `ClientMessage` interface.
+
+### Sentinel objects
+
+Similarly as in TCP/IP, where there exists a single loopback address, the protocol defines the existence of a single sentinel `ClientState` instance with the client identifier `09-localhost`. Creation of other loopback clients MUST be forbidden.
+
+Additionally, implementations **will** reserve a special connection identifier `connection-localhost` to be used by a single sentinel `ConnectionEnd` stored by default (i.e. at genesis or upgrade). The `clientIdentifier` and `counterpartyClientIdentifier` of the connection end both reference the sentinel `09-localhost` client identifier. The `counterpartyConnectionIdentifier` references the special connection identifier `connection-localhost`. The existence of a sentinel loopback connetion end enables IBC applications to establish channels directly on top of the sentinel connection. Channel handshakes can then be initiated by supplying the loopback connection identifier (`connection-localhost`) in the `connectionHops` parameter of the `ChanOpenInit` datagram.
+
+Implementations **may** also allow the creation of more connections associated with the loopback client. These connections would then have a connection identifier as generated by `generateIdentifier`.
+
+### Relayer messages
+
+Relayers supporting localhost packet flow must be adapted to submit messages from sending applications back to the originating chain.
+
+This would require first checking the underlying connection identifier on any channel-level messages. If the underlying connection identifier is `connection-localhost`, then the relayer must construct the message and send it back to the originating chain. The message MUST be constructed with a sentinel byte for the proof (`[]byte{0x01}`), since the loopback client does not need Merkle proofs of the state of a remote ledger; the proof height in the message may be zero, since it is ignored by the loopback client.
+
+Implementations **may** choose to implement loopback such that the next message in the handshake or packet flow is automatically called without relayer-driven transactions. However, implementors must take care to ensure that automatic message execution does not cause gas consumption issues.
+
+### Client initialisation
+
+Loopback client initialisation requires the latest height of the local ledger.
+
+```typescript
+function initialise(
+  height: Height): ClientState {
+    assert(height > 0)
+    return ClientState{
+      latestHeight: height
+    }
+}
+```
+
+### Validity predicate
+
+Loopback client validity checking makes sure the height is greater than zero. If the provided header is valid, the client state update logic is executed.
+
+```typescript
+function verifyClientMessage(
```

Comment:
I know that we don't use this in ibc-go, but again, I think from an abstraction point of view there's no reason why it could be done. It was ibc-go's decision to not implement localhost client this way.

Reply:
I would also remove this because it would be very inefficient for relayers to update in this way.

We should explicitly have this spec no-op or error on these methods. And have some small section explaining that the implementation should run some per-block logic to update the localhost client state height. You do not need to introduce EndBlocker or GetSelfHeight explicitly. It is up to the specific execution environment how it will be implemented